autorest/azure/async.go

```diff
@@ -0,0 +1,165 @@
+package azure
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/date"
+)
+
+const (
+	// HeaderAsyncOperation is the Azure header providing the URL from which to obtain the
+	// OperationResource for an asynchronous (aka long running) operation.
+	HeaderAsyncOperation = "Azure-AsyncOperation"
+)
+
+const (
+	// OperationCanceled says the underlying operation was canceled.
+	OperationCanceled string = "Canceled"
+	// OperationFailed says the underlying operation failed.
+	OperationFailed string = "Failed"
+	// OperationSucceeded says the underlying opertion succeeded.
+	OperationSucceeded string = "Succeeded"
+)
+
+// OperationError provides additional detail when an operation fails or is canceled.
+type OperationError struct {
+	// Code provides an invariant error code useful for troubleshooting, aggregration, and analysis.
+	Code string `json:"code,omitempty"`
+	// Message indicates what error occurred and what can be done to address the issue.
+	Message string `json:"message,omitempty"`
+}
+
+// ToError returns an error containing the code and message.
+func (oe OperationError) ToError() error {
+	return fmt.Errorf("Azure Operation Error -- code(%v), message(%v)", oe.Code, oe.Message)
+}
+
+// OperationResource defines a resource describing the state of a long-running operation.
+type OperationResource struct {
+	// Id is the identifier used in a GET for the underlying resource.
+	ID string `json:"id,omitempty"`
+	// Name matches the last segment of the Id field (typically system generated).
+	Name string `json:"name,omitempty"`
+	// Status provides the state of the operation. Non-terminal states vary by resource;
+	Status string `json:"status,omitempty"`
+	// Properties, on operation success, optionally contains per-service / per-resource values.
+	Properties map[string]interface{} `json:"properties,omitempty"`
+	// Error provides additional detail if the operation is canceled or failed.
+	Error OperationError `json:"error,omitempty"`
+	// StartTime optionally provides the time the operation started.
+	StartTime date.Time `json:"startTime,omitempty"`
+	// EndTime optionally provides the time the operation completed.
+	EndTime date.Time `json:"endTime,omitempty"`
+	// PercentComplete optionally provides the percent complete between 0 and 100.
+	PercentComplete float64 `json:"percentComplete,omitempty"`
+}
+
+// IsCanceled returns true if the operation was canceled; false otherwise.
+func (or OperationResource) IsCanceled() bool {
+	return or.Status == OperationCanceled
+}
+
+// HasFailed returns true if the operation has failed; false otherwise.
+func (or OperationResource) HasFailed() bool {
+	return or.Status == OperationFailed
+}
+
+// HasSucceeded returns true if the operation has succeeded; false otherwise.
+func (or OperationResource) HasSucceeded() bool {
```

Comment:
@ahmetalpbalkan Simplicity is good. I like the suggestion to use the switch. I'm less clear on removing the IsXxxx methods. How many symbols we're exporting does not seem to me a useful measure as there is no clear definition of what would be the right number. There are three distinct states: The two "error" states are quite different and need discrimination. Since we cannot easily discriminate with Go's builtin in error. Having the check serves a purpose. (It is better to hide string comparisons than to expose them.)

Regardless, since we're not (yet) exposing the object, I'll remove the IsXxxxx checks for now. I do not feel this is the best move.